### PR TITLE
workarounds for #1492 and #1504 and misc

### PR DESCRIPTION
**Pull request by pzinn**
- workaround for issue #1492 -- rewrote `style`. I'm still not too happy with `replace` having both a `Thing` argument and optional arguments.
A side effect is that I also had to slightly rewrite the part of `VectorGraphics` that called `style`, which means I threw in a couple of minor patches of `VectorGraphics`.
- took this opportunity to add doc for `style`.
- workaround for issue #1504 -- `htmlLiteral` now uses `regex` rather than `match`.
- minor `webapp` update.

### CODE REVIEW

File: M2/Macaulay2/packages/VectorGraphics.m2 [hunk 1]
```diff
@@ -1,7 +1,7 @@
 -- -*- coding: utf-8 -*-
 newPackage(
         "VectorGraphics",
-        Version => "0.9",
+        Version => "0.91",
```
Comment (mahrud) [inline M2/Macaulay2/packages/VectorGraphics.m2:4]: Could you also remove "A package to " from the headline, just so it matches other packages?

File: M2/Macaulay2/m2/html.m2 [hunk 2]
```diff
@@ -21,7 +21,7 @@ defaultHEAD = title -> HEAD splice { TITLE title, defaultCharset(), defaultStyle
 -----------------------------------------------------------------------------
 
 -- TODO: urlEncode
-htmlLiteral = s -> if s === null or not match("<|&|]]>|\42", s) then s else (
+htmlLiteral = s -> if s === null or regex("<|&|]]>|\42", s) === null then s else (
```
Comment (DanGrayson) [inline M2/Macaulay2/m2/html.m2:24]: Was this change needed for some reason??

**Comment by DanGrayson** [comment]
> I'm still not too happy with `replace` having both a `Thing` argument and optional arguments.

Maybe we should discuss in a separate whether to go back to `replace` taking no options, for the sake of upward compatibility.  The only option is for POSIX regular expressions.

File: M2/Macaulay2/m2/html.m2 (hunk 2 above)
Comment (mahrud) [reply M2/Macaulay2/m2/html.m2:24]: #1504.

Comment (DanGrayson) [reply M2/Macaulay2/m2/html.m2:24]: Is this change related to `lastMatch`?  It doesn't appear in the code.

Comment (pzinn) [reply M2/Macaulay2/m2/html.m2:24]: `lastMatch` is modified by `match`. `htmlLiteral` is called in `WebApp` mode all the time (to convert text to HTML). That means `lastMatch` can't be used in `WebApp` mode. If `lastMatch` becomes deprecated that won't be an issue, but for now I followed Mahrud's suggestion to replace `match` with `regex` (which does the exact same except doesn't update `lastMatch`).

**Comment by pzinn** [comment]
> > I'm still not too happy with `replace` having both a `Thing` argument and optional arguments.
> 
> Maybe we should discuss in a separate whether to go back to `replace` taking no options, for the sake of upward compatibility. The only option is for POSIX regular expressions.

sure, that's what #1492 is about.

File: M2/Macaulay2/m2/webapp.m2 [hunk 3]
```diff
@@ -3,28 +3,37 @@
...
+	)=webAppTags;
 
 texMathStart := "\\(";
 texMathEnd := "\\)";
-texWrap := x -> texMathStart | texMath x | texMathEnd; -- similar to 'tex', but only used by webapp.m2 to avoid thread-safety issues
+texWrap := x -> (  -- similar to 'tex', but only used by webapp.m2 to avoid thread-safety issues -- TODO: rewrite in a thread-safe way
+    y := texMath x;
+    if class y =!= String then error "invalid texMath output";
+    texMathStart | y | texMathEnd
+    )
```
Comment (DanGrayson) [inline M2/Macaulay2/m2/webapp.m2:24]: What aspect of this is not thread safe?

File: M2/Macaulay2/m2/webapp.m2 [hunk 4]
```diff
@@ -3,28 +3,37 @@
...
 
 htmlWithTex Thing := texWrap -- by default, we use tex (as opposed to html)
 
+stripTags := s -> replace(concatenate("[",webAppTags,"]"),"",s)
+
```
Comment (DanGrayson) [inline M2/Macaulay2/m2/webapp.m2:29]: You may wish to precompute the string `concatenate("[",webAppTags,"]")`.

File: M2/Macaulay2/packages/Text.m2 [hunk 5]
```diff
@@ -816,6 +816,16 @@ document { Key => {(validate, Hypertext),validate},
...
+document { Key => {(style, Hypertext),style},
+    Usage => "style (x,opts)",
+    Inputs => { "x" => { TO "hypertext" } },
+    Consequences => { { "Returns a new Hypertext object with its style option updated." }},
+    EXAMPLE lines ///
+    d = DIV { "Hello" }
+    d = style(d, "font-weight" => "bold")
+    peek d
+    ///
+    }
```
Comment (DanGrayson) [inline M2/Macaulay2/packages/Text.m2:828]: I guess the code for `style` could also be in the Text package.

Comment (pzinn) [reply M2/Macaulay2/packages/Text.m2:828]: Oops! Will fix that

File: M2/Macaulay2/m2/webapp.m2 (hunk 3 above)
Comment (pzinn) [reply M2/Macaulay2/m2/webapp.m2:24]: We discussed this during my previous PR (can't find the conversation for some reason); you're the one who convinced me non thread-safe code was bad. I'm merely writing a note to myself to eventually fix this; no actual change here.

File: M2/Macaulay2/packages/Text.m2 (hunk 5 above)
Comment (pzinn) [reply M2/Macaulay2/packages/Text.m2:828]: edited: oh. I thought the comment was about the documentation. let me think about it carefully before I make this change.

File: M2/Macaulay2/packages/VectorGraphics.m2 (hunk 1 above)
Comment (pzinn) [reply M2/Macaulay2/packages/VectorGraphics.m2:4]: missed that. so what's the correct formulation? just "produce SVG graphics"? I'm confused

File: M2/Macaulay2/packages/Text.m2 (hunk 5 above)
Comment (pzinn) [reply M2/Macaulay2/packages/Text.m2:828]: do you have an opinion Mahrud? since you're the one who restructured `hypertext.m2`.
In principle it would make sense to move `style` to `Text`, but it's a bit weird since at the moment `Text` is a completely empty package which only contains exports & documentation.

Comment (mahrud) [reply M2/Macaulay2/packages/Text.m2:828]: I do not.

Comment (pzinn) [reply M2/Macaulay2/packages/Text.m2:828]: OK then I'd rather not move style.

File: M2/Macaulay2/packages/VectorGraphics.m2 (hunk 1 above)
Comment (pzinn) [reply M2/Macaulay2/packages/VectorGraphics.m2:4]: I followed the `FirstPackage` template which says `an example Macaulay2 package`...
